#Exercise 10.02: Analyse the categorical variables from the Ames Housing dataset
In this exercise, we will continue our dataset exploration by analysing the categorical variables of this dataset. To do so, we will implement our own describe functions.
The dataset used for this exercise is the Ames Housing dataset compiled by Dean De Cock: http://www.amstat.org/publications/jse/v19n3/decock.pdf

1. Open on a new Colab notebook and import the pandas package

In [0]:
import pandas as pd

2. Assign the link to the AMES dataset to a variable called 'file_url':

In [0]:
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter10/dataset/ames_iowa_housing.csv'

3. Using the read_csv method from the package pandas, load the dataset into a new variable called 'df':

In [0]:
df = pd.read_csv(file_url)

4. Create a new dataframe called 'obj_df' with only the columns that are of numerical types using the method select_dtypes from pandas packages and pass in the value 'object' to the parameter 'include: 


In [0]:
obj_df = df.select_dtypes(include='object')


5. Using the attribute **columns** from pandas, extract the list of columns of this dataframe 'obj_df' and assign it to a new variable called 'obj_cols' and print its content:

Expected Output:

![Figure 27 - List of categorical variables](https://docs.google.com/uc?export=download&id=1n5gkiNhA0bPeZFUTdrXEssgqJtqwuBW0)


In [0]:
obj_cols = obj_df.columns
obj_cols

Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition'],
      dtype='object')

6. Create a function called 'describe_object' that takes a **pandas** dataframe and a column name as input parameters. Then inside the function, print out the name of the given column, its number of unique values using the method **nunique()** and the list of values and their occurence using the method **value_counts()**:

In [0]:
def describe_object(df, col_name):
  print(f"\nCOLUMN: {col_name}")
  print(f"{df[col_name].nunique()} different values")
  print(f"List of values:")
  print(df[col_name].value_counts(dropna=False, normalize=True))

7. Test this function by providing the 'df' dataframe and the column 'MSZoning':

Expected Output:

![Figure 28 - Display of the created function for the column ‘MSZoning’](https://docs.google.com/uc?export=download&id=1PE8Ive1mpHZeeC0GN5wDrjxvkaufT5Pv)


In [0]:
describe_object(df, 'MSZoning')


COLUMN: MSZoning
5 different values
List of values:
RL         0.788356
RM         0.149315
FV         0.044521
RH         0.010959
C (all)    0.006849
Name: MSZoning, dtype: float64


For the column ‘MSZoning’, the value ‘RL’ represents almost 79% of the values while ‘C (all) is only present less than 1% of the rows.

8. Create a for loop that will call the created function for every element from the list 'obj_cols':

Expected Output:

![Figure 29 - Display of the created function for the first columns contained in ‘obj_cols’](https://docs.google.com/uc?export=download&id=1cou8u_aD4Z8ec1EqDKRPFXJGGyEQUzh_)



In [0]:
for col_name in obj_cols:
  describe_object(df, col_name)


COLUMN: MSZoning
5 different values
List of values:
RL         0.788356
RM         0.149315
FV         0.044521
RH         0.010959
C (all)    0.006849
Name: MSZoning, dtype: float64

COLUMN: Street
2 different values
List of values:
Pave    0.99589
Grvl    0.00411
Name: Street, dtype: float64

COLUMN: Alley
2 different values
List of values:
NaN     0.937671
Grvl    0.034247
Pave    0.028082
Name: Alley, dtype: float64

COLUMN: LotShape
4 different values
List of values:
Reg    0.633562
IR1    0.331507
IR2    0.028082
IR3    0.006849
Name: LotShape, dtype: float64

COLUMN: LandContour
4 different values
List of values:
Lvl    0.897945
Bnk    0.043151
HLS    0.034247
Low    0.024658
Name: LandContour, dtype: float64

COLUMN: Utilities
2 different values
List of values:
AllPub    0.999315
NoSeWa    0.000685
Name: Utilities, dtype: float64

COLUMN: LotConfig
5 different values
List of values:
Inside     0.720548
Corner     0.180137
CulDSac    0.064384
FR2        0.032192
FR3        0.00

We can confirm that the column ‘Street’ is almost constant as 99.6% of the rows contains the same value: ‘Pave’. For the column ‘Alley’, almost 94% of the rows have missing values. We can also notice that 'MiscFeature' contains missing value for 96% of the rows so this column doesn't contain much information.


Excellent! We just analysed all the categorical variables from this dataset in one go. We saw how to look at the distribution of all the values contained in one feature. We also found that some of them are dominated by a single value and others have mainly missing values.